proposals/xxxx-die.md

```diff
@@ -0,0 +1,208 @@
+
+These items represent common scenarios that are fundamental to language design.
+
+### Unreachable
+
+The Swift compiler cannot detect that the following code is exhaustive. It emits an error and recommends a default clause.
+
+```swift
+switch Int.random(in: 0...2) {
+case 0: print("got 0")
+case 1: print("got 1")
+case 2: print("got 2")
+}
+
+// error: switch must be exhaustive
+// note: do you want to add a default clause?
+```
+
+That default clause should annotate the code so it's clear that the clause is unreachable and should never be run.
+
+### Not yet implemented
+
+In the `TODO` scenario, a type member is reachable but calling it should trap at runtime because the logic has not yet been implemented. 
+
+```swift
+public func semanticallyResonantMethodName() {
+    // TODO: implement this
+}
+```
+
+This call should establish why the code cannot yet be run. Using `#error` would prevent compilation and debugging. Using `#warning` might be insufficient, both in describing an unexpected failure and for "warn-as-errors" development houses.
+
+### Must override
+
+It's not uncommon to build an abstract supertype that requires methods and properties to be implemented in concrete subtypes:
+
+```swift
+class AbstractSupertype {
+    func someKeyBehavior() {
+        // Must override in subclass
+    }
+}
+```
+
+This method should never be called, as the type is abstract.
+
+### Uncallable
+
+Silencing warnings should not contradict Swift safety best practices. Some required members may add functionality that should never be called:
+
+```swift
+required init?(coder: NSCoder) {
+  super.init(coder: coder)
+  // Xcode requires unimplemented initializer
+}
+```
+
+A method that is left _deliberately_ unimplemented rather than as a `TODO` item should be annotated as such.
+
+## Proposed solution
+
+Overload the global `fatalError` function as `fatalError(because:)` and introduce `FatalReason`, a new type that vends static members to error under these universal scenarios. The precise wording and choice of each member and error message can be bikeshedded after acceptance of this proposal's concept.
+
+```swift
+/// Reasons why code should die at runtime
+public struct FatalReason: CustomStringConvertible {
+  /// Die because this code branch should be unreachable.
+  public static let unreachable = FatalReason("Should never be reached during execution.")
+  
+  /// Die because this method or function has not yet been implemented.
+  public static let notImplemented = FatalReason("Not yet implemented.")
+  
+  /// Die because a default method must be overriden by a
+  /// subtype or extension.
+  public static let mustOverride = FatalReason("Must be overriden in subtype.")
+  
+  /// Die because this functionality should never be called,
+  /// typically to silence requirements.
+  public static let uncallable = FatalReason("Should never be called.")
+  
+  /// An underlying string-based cause for a fatal error.
+  public let reason: String
```

Comment:
I'm wondering if this property is redundant, as it acts the same as a `description` property would?

```swift
public struct FatalReason: CustomStringConvertible {
  public let description: String

  public init(description: String) {
    ...
  }
}
```

Couple of ~~reasons~~ justifications for this:
- this text is a description of the `FatalReason` for the developer to see
- I would not normally expect an object to have a property with the same name. It is a bit odd to me to create a `FatalReason` with a `reason` -- because isn't the object itself representing the reason? :smile:

Reply:
I'll let @davedelong respond as this is his design.

Reply:
The current implementation was essentially a reflexive one. I'm not opposed to eliminating `let reason` and only having `let description`. That might be better from the POV that `description` is understood to _not_ be parseable or interpretable except as a debugging/diagnostic tool

Reply:
I think it is clearer that it is a description of the reason, as currently the property reads like the reason of the reason.

Perhaps you could even go so far as to make it conform to `CustomDebugStringConvertible` instead of `CustomStringConvertible ` if you want the description to be very explicitly for the developer.

Reply:
will this suggestion be taken into consideration?